src/mechanics/timeIntegration/TimeControl.cpp

```diff
@@ -0,0 +1,105 @@
+#include "TimeControl.h"
+
+
+
+void NuTo::TimeControl::Proceed()
+{
+    if (mTimeStep <= 0.0)
+        throw MechanicsException(__PRETTY_FUNCTION__,"Current timestep is 0 or negative!");
+    mPreviousTime = mCurrentTime;
+    mCurrentTime += mTimeStep;
+    if (mCurrentTime > mTimeFinal)
+    {
+        mCurrentTime = mTimeFinal;
+        mTimeStep = mTimeFinal - mCurrentTime;
+    }
+}
+
+void NuTo::TimeControl::AdjustTimestep(int iterations, int maxIterations, bool converged)
+{
+
+
+    mTimeStep = mTimeStepFunction(*this,iterations,maxIterations,converged);
+
+    if(mTimeStep > mMaxTimeStep)
+        mTimeStep =  mMaxTimeStep;
+
+    if(!converged && mPreviousTime !=mCurrentTime)
+        throw MechanicsException(__PRETTY_FUNCTION__, "No convergence with the current maximum number of "
+                                                      "iterations, either use automatic time stepping, "
+                                                      "reduce the time step or the minimal line search cut "
+                                                      "back factor. In case you provided a custom timestepping function "
+                                                      "and intend to use some kind of automatic timestepping, call "
+                                                      "the RestorePreviosTime() function of the time control before reducing "
+                                                      "the timestep.");
+}
+
+
+
+void NuTo::TimeControl::SetTimeStep(double timeStep)
+{
+    if (timeStep<=0)
+        throw MechanicsException(__PRETTY_FUNCTION__,"Timestep must be larger tham 0!");
+
+    mTimeStep = timeStep;
+}
+
+void NuTo::TimeControl::SetTimeFinal(double timeFinal)
+{
+    if (timeFinal <= mCurrentTime)
+    {
+        throw MechanicsException(__PRETTY_FUNCTION__,"Final time must be larger than current time!");
+    }
+    mTimeFinal          = timeFinal;
+}
+
+
+
+void NuTo::TimeControl::SetTimeStepFunction(std::function<double(TimeControl&,int, int, bool)> TimeStepFunction)
+{
+    mTimeStepFunction       = TimeStepFunction;
+}
+
+void NuTo::TimeControl::UseDefaultAutomaticTimestepping()
+{
+
+    SetTimeStepFunction(DefaultAutomaticTimestepFunction);
+}
+
+void NuTo::TimeControl::UseEquidistantTimestepping()
+{
+    SetTimeStepFunction([](TimeControl& rTimeControl, int iterations, int maxIterations, bool converged)->double
+                        {
+                            return rTimeControl.GetTimeStep();
+                        });
+}
+
+void NuTo::TimeControl::SetMaxTimeStep(double rMaxTimeStep)
```

Comment:
Shouldn't the max and min time step not be part of this TimeStepFunction? Since they only make sense for adaptive time stepping, whereas for constant steps this is not necessary?